### PR TITLE
Fix CronJob missed start time handling

### LINKED ISSUE #42649: Stopping cluster overnight prevents scheduled jobs from running after cluster startup.
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):
No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
CronJob
ScheduledJob
"Too many missed start times to list"
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
Feature Request

**Kubernetes version** (use `kubectl version`):

**Environment**: 
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): CentOS Linux release 7.3.1611 (Core)
- **Kernel** (e.g. `uname -a`):Linux 4.9.11-1.el7.elrepo.x86_64 #1 SMP Sat Feb 18 18:16:50 EST 2017 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: Custom 
- **Others**: 

**What happened**:
Cluster shutdown nightly to reduce costs. On startup the cluster refuses to run jobs.
The controller manager errors with:
E0307 15:40:23.754617       1 controller.go:163] Cannot determine if default/<job name> needs to be started: Too many missed start times to list

**What you expected to happen**:
The cluster will run scheduled jobs after being restored.

**How to reproduce it** (as minimally and precisely as possible):
Set a regular schedule such as `*/1 * * * *` 
Shutdown for 100 scheduled executions of the job (100 minutes).

**Anything else we need to know**:
This seems to be related to:
Function: getRecentUnmetScheduleTimes
v1.4: https://github.com/kubernetes/kubernetes/blob/release-1.4/pkg/controller/scheduledjob/utils.go#L169

### LINKED ISSUE #73169: Code related to CronJob's `.Spec.StartingDeadlineSeconds` is unreachable
**What happened**:
The variable `tooLate` can never be `true` in https://github.com/kubernetes/kubernetes/blob/27e5971c11cfcda703a39ed670a565f0f3564713/pkg/controller/cronjob/cronjob_controller.go#L264. 

This is because `getRecentUnmetScheduleTimes` never returns a scheduled time that is less than `now-StartingDeadlineSeconds`. 

In https://github.com/kubernetes/kubernetes/blob/27e5971c11cfcda703a39ed670a565f0f3564713/pkg/controller/cronjob/utils.go#L124, 

```bash
earliestTime=max(LastScheduleTime, CreationTimestamp, now-StartingDeadlineSeconds)
# It meas that, 
earliestTime >= now-StartingDeadlineSeconds
# Assume that `t` is a scheduled time that can be appended to the final result. 
earliestTime <= t < now
# So, 
now-StartingDeadlineSeconds <= t < now
```

So `tooLate` can nerver be `true` in https://github.com/kubernetes/kubernetes/blob/27e5971c11cfcda703a39ed670a565f0f3564713/pkg/controller/cronjob/cronjob_controller.go#L266. In order to make `tooLate=true`, it needs: 

```
# `t` is the scheduled time
t+StartingDeadlineSeconds<now
# Equals to 
t<now-StartingDeadlineSeconds
```

Therefore, `tooLate` can never be `true` and the next branch checking `tooLate` is unreachable. 

**What you expected to happen**:
Line 264-279 in https://github.com/kubernetes/kubernetes/blob/27e5971c11cfcda703a39ed670a565f0f3564713/pkg/controller/cronjob/cronjob_controller.go should be removed. 

**How to reproduce it (as minimally and precisely as possible)**:
`n/a`

**Anything else we need to know?**:
`n/a`

**Environment**:
- Kubernetes version (use `kubectl version`): `master`
- Cloud provider or hardware configuration: `n/a`
- OS (e.g. from /etc/os-release): `n/a`
- Kernel (e.g. `uname -a`): `n/a`
- Install tools: `n/a`
- Others: `n/a`

@kubernetes/sig-scheduling-bugs

### PR DESCRIPTION
**Pull request by praseodym**
**What type of PR is this?**

/kind bug
/sig apps
/area workload-api/cronjob
/priority important-soon

**What this PR does / why we need it**:

This removes the arbitrary limit of 100 missed start times and will instead always schedule a job when an execution was missed and the `Spec.StartingDeadlineSeconds` period has not yet passed (if set).

To prevent starting multiple jobs if multiple start times were missed, `Status.LastScheduleTime` is now set to the actual start time of the job instead of the original scheduled start time.

The "missed starting window" warning message is removed because it was broken (see kubernetes/kubernetes#73169).

**Which issue(s) this PR fixes**:

Fixes #42649
Fixes #73169

**Special notes for your reviewer**:

See #42649 for numerous examples of problems caused by the current arbitrary limit of 100 missed start times, which is easily hit in real-world usage. For example, with a cron job set to execute every minute, 101 minutes of controller downtime (due to maintenance or some bug in the controller) already causes the cron job to permanently fail.

**Does this PR introduce a user-facing change?**:

```release-note
Fix CronJob missed start time handling: arbitrary limit of 100 missed start times is removed. CronJob `Status.LastScheduleTime` is now set to the actual start time of the job instead of the original scheduled start time.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
- [Usage]: https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/
```
The above doc page will need to be updated if this PR is merged, I will create a PR in https://github.com/kubernetes/website once this is approved.

### CODE REVIEW

**Comment by kow3ns** [comment]
It's not actually clear to me that we should do this.

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go]: I'd prefer something more concrete in the docs than just `some time has passed`.

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go]: `Sub` is the method you should be using here.

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go]: Controller cannot ~~to~~ schedule

File: pkg/controller/cronjob/utils.go
Comment (soltysh) [inline pkg/controller/cronjob/utils.go]: This is wrong and the previous logic was trying to prevent us from doing this. Imagine the situation described in the comment you removed. 5pm Friday something is borked, you only notice a problem with it on Tuesday, in the mean time 80 jobs were hold up, now suddenly we'll start creating all of those 80 jobs, which is wrong. If you want to fix it, I'd prefer we emit an event that there was a problem and we missed X starts and pick only the last date. The current logic does not have that protection.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/controller/cronjob/utils.go
Comment (praseodym) [reply pkg/controller/cronjob/utils.go]: This PR will not work in the way you describe because we now set `Status.LastScheduleTime` to the actual start time instead of the original scheduled time. 

In your example, if we consider this an hourly crib job: the Friday 6pm will get scheduled after fixing the problem, let’s say Tuesday 10:30am, but the next run after that will be Tuesday 11am. Other jobs will be skipped.

The tests have been changed to reflect this behaviour as well.

**Comment by danbeaulieu** [comment]
I do not have an opinion on the implementation but I believe this will fix behavior in CronJobs that is surprising. 

If I have a cronjob scheduled to run every 10 minutes, and I have some downtime in my cluster, when it comes back I expect it to start the job every 10 minutes again. I do not expect it to refuse to schedule the jobs ever again since there was downtime, which is the current behavior.

**Comment by fredsted** [comment]
Thanks for this PR!

As it stands now, Kubernetes cronjobs are totally unpredictable. 

We've went so far as to build a custom cronjob monitoring tool to alert us when they go into the "FailedNeedsStart" state, but we still run into discovering that a cronjob that *should* have been running isn't for some reason.

The current functionality does not match the expectation for the word "cronjob". What is expected is that cronjobs are maybe not always run, but the system *will continue to try running them forever*, and not give up after a totally arbitrary limit of 100.

File: pkg/controller/cronjob/utils.go
Comment (bbbmj) [reply pkg/controller/cronjob/utils.go]: > Other missed executions will be skipped.

Just skip and go on is not a good practice, it may result in unpredictable situations.

Make missed starts configurable maybe better.

Comment (aermakov-zalando) [reply pkg/controller/cronjob/utils.go]: @bbbmj It's still better than not running the cronjob at all unless it's deleted and recreated (which is what happens now), so making this configurable should be done as a separate PR.

Comment (soltysh) [reply pkg/controller/cronjob/utils.go]: @praseodym I see your point, apologies for the initial misunderstanding. I see a lot of benefit in your approach, but at the same time this fix will break backwards compatibility. We need to ensure that previous behaviour is somehow configurable, maybe through additional field (iow. how many missed jobs can be ignored, or whether a cronjob should be paused after significant failure). Can we start with the latter, which seems to be the easiest and is both backwards compatible and at the same time will make users more in control of getting cronjob back to shape? I've added this topic to sig-apps agenda for the next meeting (according to my calendar it should be Jan 13th). Let's discuss the best possible approach there and we can continue from there.

Comment (praseodym) [reply pkg/controller/cronjob/utils.go]: I’d be against introducing additional complexity in the codebase, whereas this PR (in my opinion) is currently mostly cleanup instead (see e.g. #73169, which this PR also fixes).

Furthermore, I don’t see anyone relying on the current behaviour of CronJobs breaking after 100 missed executions (they’re not paused — they break and need to be completely recreated). The discussion in #42649 also doesn’t contain any comments defending the current behaviour (it does contain references to [workarounds that should make any Kubernetes maintainer cry](https://github.com/zalando-incubator/kubernetes-on-aws/pull/2734)). Finally, the [KEP for Graduating CronJob to GA](https://github.com/kubernetes/enhancements/pull/978/files#diff-4128ed7a115ba561a9c52e3778b45a46R113) (unfortunately not merged) also lists this issue as a bug, not as an enhancement.

I propose we move forward with this PR as-is, and include a clear release note instead.

**Comment by neolit123** [comment]
@kubernetes/sig-apps-pr-reviews

**Comment by barney-s** [comment]
> Thanks for this PR!
> 
> As it stands now, Kubernetes cronjobs are totally unpredictable and should not be used.
> 
> We've went so far as to build a custom cronjob monitoring tool to alert us when they go into the "FailedNeedsStart" state, but we still run into discovering that a cronjob that _should_ have been running isn't for some reason.
> 
> The current functionality does not match the expectation for the word "cronjob". What is expected is that cronjobs are maybe not always run, but the system _will continue to try running them forever_, and not give up after a totally arbitrary limit of 100.

@fredsted  Please review the cronjob GA KEP https://github.com/kubernetes/enhancements/pull/978
and provide your inputs there. We have time to fix the cronjob behavior before we graduate to stable.

File: pkg/controller/cronjob/utils.go
Comment (vllry) [inline pkg/controller/cronjob/utils.go]: Have you benchmarked the worst-case scenarios for removing this?

Given this specific code, the limit exists for a reason. The cost is small but nontrivial to iterate over a long time window (say, a minutely cron that has failed for a year), and it really adds up with the current synchronous CronJob processing.

There are ways to avoid traversing the entire time window, and thereby avoiding the need for this cap: https://github.com/kubernetes/enhancements/pull/978#discussion_r393457336
I was planning on open-sourcing a stopgap solution (example 2 in that comment),
until a better solution is implimented in the CronJob GA work.

cc @yangkev

Comment (praseodym) [reply pkg/controller/cronjob/utils.go]: Thanks for your comment!  This PR actually removes the entire `for` loop, i.e. there is no iteration to be limited any more. The split diff makes this a lot easier to see than the default unified diff.

File: pkg/controller/cronjob/cronjob_controller.go
Comment (praseodym) [reply pkg/controller/cronjob/cronjob_controller.go]: I agree that that functionality might be useful, but it’s currently broken (https://github.com/kubernetes/kubernetes/issues/73169). I would prefer not to add any more complexity to this PR, which has been open for over 7 months now, and get it merged instead. We can then do a followup PR to get the “Deadline missed” event back in.

**Comment by mbelang** [comment]
Will that fix be back-ported to prior version of k8s e.g. 1.14?

**Comment by praseodym** [comment]
> Will that fix be back-ported to prior version of k8s e.g. 1.14?

[Kubernetes 1.14 is unsupported](https://github.com/kubernetes/sig-release/blob/master/releases/patch-releases.md#114-and-older), so unless someone else (e.g. your cloud provider) backports it, this will not be fixed in 1.14. Once a fix gets merged, it _might_ get cherry-picked into Kubernetes versions that are actively supported (currently 1.16, 1.17 and 1.18).

**Comment by niemyjski** [comment]
Any idea on the status of this?

**Comment by praseodym** [comment]
Rebased

**Comment by 2rs2ts** [comment]
I believe the site will also need to be updated for this https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#cron-job-limitations

**Comment by praseodym** [comment]
@2rs2ts that's correct, as mentioned in the initial PR message I will do this as soon as this is merged. If there's another procedure for this please let me know!

**Comment by 2rs2ts** [comment]
@praseodym maybe I'm misreading but there are two docs I linked and your summary only links one of them. Maybe I should have said "this doc also" but I was just listing the places where it occurs to be complete

**Comment by Guldanx** [comment]
@praseodym @soltysh Can we go ahead and merge this PR? Or is there something else blocking us?

**Comment by praseodym** [comment]
> @praseodym @soltysh Can we go ahead and merge this PR? Or is there something else blocking us?

This PR still needs to be approved by a reviewer before it can be merged.

**Comment by 2rs2ts** [comment]
Does #93370 obsolete this PR? I am confused about it.

**Comment by praseodym** [comment]
> Does #93370 obsolete this PR? I am confused about it.

That PR introduces a new v2 controller, but still includes the 100 missed start times limit 'feature' from the v1 controller.

**Comment by soltysh** [comment]
> That PR introduces a new v2 controller, but still includes the 100 missed start times limit 'feature' from the v1 controller.

This is not true at this point.
/close

**Comment by unixfox** [comment]
Why is this got closed?

**Comment by 2rs2ts** [comment]
Sounds like they're expecting people to upgrade to the v2 cronjob controller to get the fix? Seems a bit rough

**Comment by b0b0haha** [comment]
> Sounds like they're expecting people to upgrade to the v2 cronjob controller to get the fix? Seems a bit rough
> 
> 听起来他们期待人们升级到 v2 cronjob 控制器来获得修复? 似乎有点粗糙

Could you please point that why v2 cronjob controller fix the issue? It seems confocus me.

**Comment by b0b0haha** [comment]
> Sounds like they're expecting people to upgrade to the v2 cronjob controller to get the fix? Seems a bit rough

Could you please point out why the v2 cronjob controller can get the fix? It still includes the 100 missed start times limit 'feature' from the v1 controller.
https://github.com/alaypatel07/kubernetes/blob/8d7dd4415e28bded77667ce857e1c58016f9ab3a/pkg/controller/cronjob/utils.go#L194-L214

**Comment by 2rs2ts** [comment]
@b0b0haha I don't know, that was over a year ago. Maybe there had been discussion somewhere else about reimplementing this fix, but for the v2 controller, but it sounds like that never happened.

I see comments from people thinking that the "feature" had been fixed (i.e. removed) in the new implementation, but maybe that is mistaken: https://github.com/kubernetes/kubernetes/issues/42649#issuecomment-876474787

**Pull request closed without merging**
